crates/provider-sdk/src/provider.rs

```diff
@@ -0,0 +1,541 @@
+                                        ).in_current_span().await {
+                                            error!(error = %e, "unable to publish invocation response error");
+                                        }
+                                    }
+                                }
+                            };
+                        }.instrument(span)); /* spawn */
+                    } /* next */
+                }
+            } /* loop */
+        });
+        Ok(handle)
+    }
+
+    async fn handle_rpc<P>(
+        &self,
+        provider: P,
+        inv: Invocation,
+    ) -> Result<Vec<u8>, ProviderInvocationError>
+    where
+        P: Provider + Clone,
+    {
+        let inv = self.rpc_client.dechunk(inv).await?;
+        let (inv, claims) = self
+            .rpc_client
+            .validate_invocation(inv)
+            .await
+            .map_err(InvocationError::from)?;
+        self.validate_provider_invocation(&inv, &claims)
+            .await
+            .map_err(InvocationError::from)?;
+        let span = tracing::debug_span!("dispatch", public_key = %inv.origin.public_key, method = %inv.operation);
+        provider
+            .dispatch(
+                Context {
+                    actor: Some(inv.origin.public_key.clone()),
+                    tracing: inv.trace_context.into_iter().collect(),
+                },
+                inv.operation,
+                Cow::Owned(inv.msg),
+            )
+            .instrument(span)
+            .await
+    }
+
+    async fn subscribe_shutdown<P>(
+        &self,
+        provider: P,
+        shutdown_tx: tokio::sync::broadcast::Sender<bool>,
+    ) -> ProviderResult<JoinHandle<()>>
+    where
+        P: Provider,
+    {
+        let shutdown_topic = format!(
+            "wasmbus.rpc.{}.{}.{}.shutdown",
+            &self.lattice_prefix, &self.host_data.provider_key, self.host_data.link_name
+        );
+        debug!("subscribing for shutdown : {}", &shutdown_topic);
+        let mut sub = self.rpc_client.client().subscribe(shutdown_topic).await?;
+        let rpc_client = self.rpc_client.clone();
+        let host_id = self.host_data.host_id.clone();
+        let handle = tokio::spawn(async move {
+            loop {
+                let msg = sub.next().await;
+                // Check if we really need to shut down
+                if let Some(async_nats::Message {
+                    reply: Some(reply_to),
+                    payload,
+                    ..
+                }) = msg
+                {
+                    let shutmsg: ShutdownMessage = serde_json::from_slice(&payload).unwrap_or_default();
+                    // Backwards compatibility - if no host (or payload) is supplied, default
+                    // to shutting down unconditionally
+                    if shutmsg.host_id.is_empty() {
+                        warn!("Please upgrade your wasmcloud host to >= 0.59.0 if you use lattices with multiple hosts.")
```

Comment:
It might be time to assume all hosts will provide a host ID on shutdown messages

Reply:
Oh right. I tried to remove most of the old stuff. Let me go remove this